### PR TITLE
fix: replace cron with systemd for demo server (prevents process leak)

### PR DESCRIPTION
**Pull request by julienld**
## Summary

- **Root cause**: the weekly cron job spawned a new \`hamcp-test-env\` Python process every Monday without killing the old one. After 34 days, 65+ zombie processes had accumulated on the GCP demo server.
- **Fix**: replace cron entirely with a systemd service (\`hamcp-demo.service\`) + daily timer (\`hamcp-demo-update.timer\`). systemd ensures exactly one instance runs, restarts it on failure, and the daily update runs as root with \`sudo -i -u\` for the git pull then \`systemctl restart hamcp-demo\`.
- **Migration**: setup.sh strips old cron entries, kills leaked processes via \`pkill\`, and removes any legacy sudoers rule on re-run.

## What was broken on the server

- 65 leaked processes dating to 2026-03-30 (SIGKILL required — processes were owned by a different user than the SSH user)
- Crontab had been cleared, so \`@reboot\` entry was gone — a reboot would have left HA permanently down
- TLS certificate had expired (Caddy couldn't renew while previous process held state) — fixed by restarting Caddy
- Demo HA was accessible but the process pile would eventually cause OOM

## Changes

- **systemd service** (\`hamcp-demo.service\`): \`KillSignal=SIGINT\` so Python's cleanup handler runs on stop/restart; \`Wants=network-online.target\` for reliable cold-boot startup
- **Daily update timer** (\`hamcp-demo-update.timer\`): replaced weekly Monday schedule with daily 3am; runs as root, uses \`sudo -i -u\` for git pull to ensure correct \`$HOME\`; \`docker image prune -af\` prevents disk fill
- **Crontab migration**: guarded with \`grep -q\` (no empty crontab on fresh hosts); \`grep -v\` wrapped in \`{ ... || true; }\` to survive \`set -euo pipefail\` when all lines match
- **Cleanup order**: \`systemctl stop\` before \`pkill\` so systemd doesn't race its own auto-restart
- **uv install**: switched to \`sudo -i -u\` so \`\$HOME\` is correct regardless of host \`always_set_home\` setting
- **README**: updated timing (3–10 min typical, up to 15 on first pull), full-reset snippet uses ancestor filter instead of stopping all containers

## Applied live on the GCP demo server

All changes verified on \`instance-20251202-013322\` through multiple re-runs:
- 65 zombie processes killed
- systemd service + daily timer created, enabled, running clean (2 processes, 1 container)
- TLS certificate renewed
- Demo HA accessible and healthy at ha-mcp-demo-server.qc-h.net

## Tested

Fresh uninstall → reinstall cycle verified: service starts, timer enabled, sudoers rule absent, HTTP+HTTPS both return 200.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/lab-setup/setup-ha-mcp.sh [hunk 1]
```diff
@@ -82,15 +82,69 @@ else
...
+User=${SETUP_USER}
+Group=${SETUP_USER}
+WorkingDirectory=${SETUP_HOME}/ha-mcp
+Environment=HA_TEST_PORT=${HA_PORT}
+ExecStart=${UV_PATH} run hamcp-test-env --no-interactive
+Restart=on-failure
+RestartSec=30s
+StandardOutput=append:/var/log/hamcp-demo.log
+StandardError=append:/var/log/hamcp-demo.log
+TimeoutStopSec=60
```
[the remark this answers is not in the document]
Comment (julienld) [reply tests/lab-setup/setup-ha-mcp.sh]: ✅ Fixed in ac75d93. Removed `StandardOutput=append:/var/log/hamcp-demo.log` and `StandardError=append:/var/log/hamcp-demo.log` from both services. All output now goes to the systemd journal (`journalctl -u hamcp-demo`). README and success output updated to remove the log file references.

File: tests/lab-setup/setup-ha-mcp.sh [hunk 2]
```diff
@@ -146,19 +200,20 @@ APT::Periodic::AutocleanInterval "7";
 AUTOEOF
 
 #=============================================================================
-# 9. STOP OLD CONTAINERS
-info "Cleaning up old containers..."
+# 9. STOP OLD CONTAINERS + PROCESSES
+info "Cleaning up old containers and processes..."
 docker ps -aq --filter "ancestor=ghcr.io/home-assistant/home-assistant" | xargs -r docker rm -f 2>/dev/null || true
+docker ps -aq --filter "ancestor=testcontainers/ryuk" | xargs -r docker rm -f 2>/dev/null || true
```
[the remark this answers is not in the document]
Comment (julienld) [reply tests/lab-setup/setup-ha-mcp.sh:209]: ✅ Fixed in ac75d93. Added `pkill -u "$SETUP_USER" -f "hamcp-test-env" 2>/dev/null || true` to the cleanup step alongside the docker container removal.

File: tests/lab-setup/setup-ha-mcp.sh [hunk 3]
```diff
@@ -183,11 +238,12 @@ if docker ps | grep -q "home-assistant"; then
     echo "Credentials: dev / dev"
     echo ""
     echo "Logs:        docker logs -f $CONTAINER"
-    echo "Startup log: tail -f /tmp/hamcp.log"
+    echo "Service log: journalctl -u hamcp-demo -f"
+    echo "Full log:    tail -f /var/log/hamcp-demo.log"
```
[the remark this answers is not in the document]
Comment (julienld) [reply tests/lab-setup/setup-ha-mcp.sh]: ✅ Fixed in ac75d93. Removed the log file line from the success output (and from the README).

File: tests/lab-setup/setup-ha-mcp.sh [hunk 4]
```diff
@@ -82,15 +82,69 @@ else
...
+
+[Service]
+Type=oneshot
+User=${SETUP_USER}
+Group=${SETUP_USER}
+WorkingDirectory=${SETUP_HOME}/ha-mcp
+ExecStart=/bin/bash -c 'git pull --ff-only'
+ExecStartPost=/bin/systemctl restart hamcp-demo
+StandardOutput=append:/var/log/hamcp-demo.log
+StandardError=append:/var/log/hamcp-demo.log
```
[the remark this answers is not in the document]
Comment (julienld) [reply tests/lab-setup/setup-ha-mcp.sh]: ✅ Fixed in ac75d93. Update service now runs as root with `sudo -u ${SETUP_USER}` for the git pull, so `ExecStartPost=/usr/bin/systemctl restart hamcp-demo` works without any sudoers rule. Also added `After=network-online.target` + `Wants=network-online.target` to the update service.

File: tests/lab-setup/README.md [hunk 5]
```diff
@@ -2,95 +2,110 @@
...
-- Prunes unused Docker images (prevents disk fill)
-- Restarts hamcp-test-env with fresh container
+1. **Swap** — Creates 6GB swap for small VMs
+2. **Packages** — Installs curl, git, ca-certificates, gnupg
+3. **Docker** — Via official get.docker.com script
+4. **uv** — Python package manager for running ha-mcp
+5. **ha-mcp repo** — Clones to `~/ha-mcp` (or pulls if it already exists)
+6. **Systemd service** — Creates `hamcp-demo.service`: starts on boot, restarts automatically on failure
+7. **Systemd timer** — Creates `hamcp-demo-update.timer`: pulls latest code and restarts the service every Monday at 3am
+8. **Sudoers rule** — Allows the setup user to run `systemctl restart hamcp-demo` without a password (needed by the weekly timer)
```
[the remark this answers is not in the document]
Comment (julienld) [reply tests/lab-setup/README.md]: ✅ Fixed in da8ed5d. Removed the sudoers rule step from the README and renumbered the remaining steps.

File: tests/lab-setup/setup-ha-mcp.sh [hunk 6]
```diff
@@ -82,15 +83,66 @@ else
...
+
+cat > /etc/systemd/system/hamcp-demo-update.service << SVCEOF
+[Unit]
+Description=HA-MCP Demo Weekly Update
+After=network-online.target docker.service
+Wants=network-online.target
+
+[Service]
+Type=oneshot
+ExecStart=/usr/bin/sudo -u ${SETUP_USER} /usr/bin/git -C ${SETUP_HOME}/ha-mcp pull --ff-only
```
[the remark this answers is not in the document]
Comment (julienld) [reply tests/lab-setup/setup-ha-mcp.sh]: ✅ Fixed in da8ed5d. Added `ExecStart=/usr/bin/docker image prune -af` as a second ExecStart line (valid for Type=oneshot, runs sequentially). Also switched to `sudo -i -u` for the git pull to ensure $HOME is correctly set.

File: tests/lab-setup/setup-ha-mcp.sh [hunk 7]
```diff
@@ -82,15 +83,66 @@ else
...
+ExecStartPost=/usr/bin/systemctl restart hamcp-demo
+SVCEOF
+
+cat > /etc/systemd/system/hamcp-demo-update.timer << SVCEOF
+[Unit]
+Description=HA-MCP Demo Weekly Update Timer
+Requires=hamcp-demo-update.service
```
[the remark this answers is not in the document]
Comment (julienld) [reply tests/lab-setup/setup-ha-mcp.sh]: ✅ Fixed in da8ed5d. Removed `Requires=hamcp-demo-update.service` from the timer unit — the timer triggers the matching service by name automatically.

**Review by Patch76 — COMMENTED** [review]
Codex's P1 on `tests/lab-setup/setup-ha-mcp.sh:123` is correct in primary sources, and worth a small fix in this PR.

`tests/test_env_manager.py` registers no `signal.signal(SIGTERM, ...)` handler and no `atexit` — the only catches are `except KeyboardInterrupt` (lines 226, 310, 326). The cleanup that calls `env.stop_container()` lives in the `finally:` at the bottom of `main()`. Python's default SIGTERM handling terminates the process without running `finally`, so when `hamcp-demo-update.service` fires `ExecStartPost=/usr/bin/systemctl restart hamcp-demo` weekly, systemd sends the default SIGTERM and the testcontainers HA container is not stopped by the Python process. The next `Restart=on-failure` start then races against the orphaned container's bind on the fixed `HA_TEST_PORT=8123` (`Environment=` line 103, `with_bind_ports(8123, port)` at `test_env_manager.py:114`).

Smallest patch is in the unit file rather than touching Python:

```ini
[Service]
KillSignal=SIGINT
TimeoutStopSec=30
```

`KillSignal=` is inherited by `RestartKillSignal=` for `systemctl restart`, SIGINT becomes `KeyboardInterrupt` in Python, and the existing `finally` path runs `stop_container()` cleanly. A Python-side SIGTERM handler would also work if you'd rather keep the unit minimal.

One migration nit on line 90:

```sh
sudo -u "$SETUP_USER" crontab -l 2>/dev/null | grep -v "hamcp-test-env" | sudo -u "$SETUP_USER" crontab - 2>/dev/null || true
```

For a user with no prior crontab, `crontab -l` exits 1 with empty stdout, `grep -v` outputs nothing, and `crontab -` then installs an empty crontab for that user — a small side-effect on fresh hosts. Wrapping the rewrite in `if sudo -u "$SETUP_USER" crontab -l 2>/dev/null | grep -q "hamcp-test-env"; then ... fi` keeps it idempotent for both fresh and migrating systems.

File: tests/lab-setup/setup-ha-mcp.sh [hunk 8]
```diff
@@ -82,15 +83,66 @@ else
...
+[Unit]
+Description=HA-MCP Demo Weekly Update
+After=network-online.target docker.service
+Wants=network-online.target
+
+[Service]
+Type=oneshot
+ExecStart=/usr/bin/sudo -i -u ${SETUP_USER} /usr/bin/git -C ${SETUP_HOME}/ha-mcp pull --ff-only
+ExecStart=/usr/bin/docker image prune -af
+ExecStartPost=/usr/bin/systemctl restart hamcp-demo
```
[the remark this answers is not in the document]
Comment (julienld) [reply tests/lab-setup/setup-ha-mcp.sh:127]: ✅ Fixed in 7d8f101. Added `KillSignal=SIGINT` to `hamcp-demo.service` — systemd will now send SIGINT instead of SIGTERM, which Python converts to `KeyboardInterrupt`, triggering the existing cleanup path in `hamcp-test-env`. No changes needed to the Python source.

**Review by julienld — COMMENTED** [review]
✅ Both points addressed:

**KillSignal=SIGINT** (7d8f101): Already applied before this review landed — same root cause analysis. The `KillSignal=` is inherited by `RestartKillSignal=` so it covers both manual and timer-triggered restarts. Kept `TimeoutStopSec=60` since HA shutdown can take a moment on slow VMs.

**Crontab migration nit** (56b37c4): Fixed — now only rewrites the crontab when `hamcp-test-env` entries are actually present, so fresh hosts are unaffected.

**Comment by Patch76** [comment]
A few non-blocking follow-up notes flagged for visibility — none affect the merge decision; both prior asks (`KillSignal=SIGINT`, crontab grep gate) are addressed at HEAD.

**1. Main service is missing `Wants=network-online.target`** (`setup-ha-mcp.sh:95-114`)

The main `hamcp-demo.service` block has `After=docker.service network-online.target` and `Requires=docker.service`, but no `Wants=network-online.target`. `After=` alone establishes ordering but does not pull the target — if no other unit on the host pulls `network-online.target`, the main service can start before network is up. On first cold boot when Docker has to pull the ~600 MB HA image, a network-not-ready race is possible. `Restart=on-failure` + `RestartSec=30s` recovers, but the update service (which has both `After=` AND `Wants=` at `:119-120`) is the right shape for the main service too.

**2. `pkill` runs before the explicit `systemctl restart`** (`setup-ha-mcp.sh:208,213`)

On a re-run where systemd is already managing the service, `pkill -u "$SETUP_USER" -f "hamcp-test-env"` at `:208` kills the unit's child; systemd then schedules its own auto-restart per `Restart=on-failure`; the explicit `systemctl restart` at `:213` forces immediate restart. The unit ends up fighting itself for a few seconds. Cleaner shape would be `systemctl stop hamcp-demo 2>/dev/null || true` BEFORE the pkill, so the unit isn't competing with the cleanup.

**3. Pre-existing latent: uv installer may install to `/root/.local/bin`** (`setup-ha-mcp.sh:70`)

```sh
sudo -u "$SETUP_USER" bash -c 'curl -LsSf https://astral.sh/uv/install.sh | sh'
```

`sudo -u` without `-i` doesn't always reset `$HOME` — depends on sudo's `always_set_home` setting. On hosts where it's off, `$HOME` stays at `/root`, uv's installer puts the binary at `/root/.local/bin/uv`, while the `[[ ! -f "$UV_PATH" ]]` guard nearby checks `$SETUP_HOME/.local/bin/uv` — the mismatch would cause a re-install on every script run.

This is pre-existing in master, not introduced by this PR. Same fix as the `:124` git-pull update (`sudo -i -u`) would close it. Worth a separate follow-up issue rather than adding scope here.

**Review by Patch76 — DISMISSED** [review]
Approve. Both prior asks are addressed at HEAD `56b37c4`:

- `KillSignal=SIGINT` (+ `TimeoutStopSec=60`) on `hamcp-demo.service` — SIGINT reaches the existing `KeyboardInterrupt`/`finally:` cleanup path in `tests/test_env_manager.py`, so `systemctl stop`/`restart` no longer leaks the testcontainers HA container.
- Crontab migration on fresh hosts is now idempotent — gated on `grep -q "hamcp-test-env"`, so hosts without a prior crontab don't get an empty one installed as side effect.

Late commits (`ac75d93` Gemini round 1, `da8ed5d` Gemini round 2, `becaa4d` image-column race fix, `7d8f101` SIGINT, `56b37c4` crontab gate) look clean. README + units are internally consistent. CI green.

Posted three small non-blocking follow-up notes in a separate comment for visibility — none affect the merge decision.

**Comment by julienld** [comment]
✅ All three points addressed:

**1. `Wants=network-online.target` on main service** (b469c1d) — added. `After=` alone establishes ordering but doesn't pull the target; `Wants=` makes the unit declare the dependency so it's always activated on cold boot.

**2. `systemctl stop` before `pkill`** (b469c1d) — added as the first cleanup action. On re-runs where systemd manages the service, this lets the unit shut down cleanly (with `KillSignal=SIGINT` triggering the cleanup handler) before we remove containers and kill stray processes.

**3. `sudo -u` → `sudo -i -u` for uv install** (bfc585a) — fixed. Login shell ensures `$HOME` is set to the target user's home regardless of the host's `always_set_home` sudo setting. Also closed #1142.

**Review by Patch76 — COMMENTED** [review]
Re-review of HEAD `bfc585a`. systemd story holds up: dependency order, `Type=oneshot` chain, `KillSignal=SIGINT`, step-9 cleanup. Heredoc quoting is correct in both directions.

One pipefail edge case in the cron migration plus two README↔script mismatches. Inline.

File: tests/lab-setup/setup-ha-mcp.sh [hunk 9]
```diff
@@ -82,15 +83,70 @@ else
...
-    sudo -u "$SETUP_USER" crontab -l 2>/dev/null | grep -v "hamcp-test-env" || true
-    echo "$CRON_REBOOT"
-    echo "$CRON_WEEKLY"
-) | sudo -u "$SETUP_USER" crontab -
+# 6. SYSTEMD SERVICE (replaces cron - ensures only one instance runs, auto-restarts)
+info "Setting up systemd service..."
+
+# Remove old cron entries if they exist (migration from cron-based setup)
+if sudo -u "$SETUP_USER" crontab -l 2>/dev/null | grep -q "hamcp-test-env"; then
+    sudo -u "$SETUP_USER" crontab -l 2>/dev/null | grep -v "hamcp-test-env" | sudo -u "$SETUP_USER" crontab -
```
Comment (Patch76) [inline tests/lab-setup/setup-ha-mcp.sh]: Under `set -euo pipefail` (line 2), this pipeline aborts mid-migration when the legacy crontab consists *only* of `hamcp-test-env` lines — exactly the realistic case on hosts created by the previous cron-based version of this script. `grep -v "hamcp-test-env"` has no non-matching lines to emit, exits 1, and pipefail propagates. The `if grep -q "hamcp-test-env"` guard on line 90 confirms a match exists but says nothing about whether non-matching lines also exist.

Reproduced locally (Bash 5.x):

```sh
$ bash -c 'set -euo pipefail; echo "hamcp-test-env" | grep -v "hamcp-test-env" | cat'; echo rc=$?
rc=1
```

Smallest fix: `… | { grep -v "hamcp-test-env" || true; } | …`. Or capture to a tempfile / `crontab -r` when the filtered output is empty.

This regresses commit `56b37c4`'s "fresh-host" guard for a different host class — legacy hosts with only the cron entries, which are precisely the ones this migration targets.

File: tests/lab-setup/README.md [hunk 10]
```diff
@@ -2,95 +2,108 @@
...
 sudo ./setup-ha-mcp.sh your-domain.example.com
 ```
 
+> **Important:** Use `sudo ./setup-ha-mcp.sh`, not `sudo su -` followed by `./setup-ha-mcp.sh`.
+> The script uses `$SUDO_USER` to identify which user to configure. Running as root directly loses that information and the script will exit with an error.
+>
+> If calling from a **root cron job**, set it explicitly:
+> `SUDO_USER=youruser ./setup-ha-mcp.sh your-domain.example.com`
+
+Setup takes about 3–10 minutes (depending on image pull speed). Home Assistant starts automatically.
```
Comment (Patch76) [inline tests/lab-setup/README.md]: The HA-ready wait runs `for i in {1..180}` with `sleep 5` (≈ 15 min — see the justifying comment at `setup-ha-mcp.sh:218–220`). README says "3–10 minutes". Fine for the warm-image case, but on a fresh GCP image with a cold pull users wait longer than the README implies. Either bump to "3–15" or qualify as "3–10 typical, up to 15 on slow links".

File: tests/lab-setup/README.md [hunk 11]
```diff
@@ -2,95 +2,108 @@
...
-docker stop $(docker ps --filter "ancestor=ghcr.io/home-assistant/home-assistant" -q)
+sudo systemctl restart hamcp-demo
+```
 
-# Start
-cd ~/ha-mcp && HA_TEST_PORT=8123 ~/.local/bin/uv run hamcp-test-env --no-interactive &
+### Full reset (re-clone and restart from scratch)
+```bash
+sudo systemctl stop hamcp-demo
+docker stop $(docker ps -q) 2>/dev/null; docker rm $(docker ps -aq) 2>/dev/null
```
Comment (Patch76) [inline tests/lab-setup/README.md]: This stops/removes *every* container on the host, not just HA/hamcp ones. Step 9 in the script filters by `--filter ancestor=ghcr.io/home-assistant/home-assistant` (plus the testcontainers ryuk image) — the README's "full reset" block could mirror that to stay safe on hosts with unrelated workloads. On a single-purpose demo VM the impact is nil, but the snippet gets copy-pasted.

File: tests/lab-setup/setup-ha-mcp.sh (hunk 9 above)
Comment (julienld) [reply tests/lab-setup/setup-ha-mcp.sh]: ✅ Fixed in 1116101. Wrapped the `grep -v` in `{ grep -v "hamcp-test-env" || true; }` so `pipefail` doesn't propagate when the legacy crontab is entirely hamcp-test-env lines and grep exits 1 with empty output.

File: tests/lab-setup/README.md (hunk 10 above)
Comment (julienld) [reply tests/lab-setup/README.md]: ✅ Fixed in 1116101. README now reads: '3–10 minutes on subsequent runs. Allow up to 15 minutes on first install when the ~600 MB HA image must be pulled.'

File: tests/lab-setup/README.md (hunk 11 above)
Comment (julienld) [reply tests/lab-setup/README.md]: ✅ Fixed in 1116101. Full reset snippet now uses `docker ps -aq --filter ancestor=...` for both the HA and ryuk images, matching what step 9 of the script does — safe to copy-paste on shared hosts.
